### DISCUSSION TITLE
HOW? FastAPI + Flet

### DISCUSSION THREAD

**Discussion by RRozi** [flet-dev/flet#3936] [Q&A]
### Question

How can I get telegram api data and then work with it in frontend flet?

### Code sample

```python
import asyncio
from contextlib import asynccontextmanager

import requests
from fastapi import FastAPI, Request
from fastapi.responses import HTMLResponse, RedirectResponse

import flet as ft
import flet.fastapi as flet_fastapi

import api
import gui


@asynccontextmanager
async def lifespan(app: FastAPI):
    await flet_fastapi.app_manager.start()
    yield
    await flet_fastapi.app_manager.shutdown()


app = FastAPI(lifespan=lifespan)


datt = {}

async def root_main(page: ft.Page):

    page.add(ft.Text("This is root app!"))
    #  Get data -> work on the front end

@app.post("/get_data")
async def send_name(request: Request):
    data = await request.json()
    print(data)

@app.get("/data")
async def read_root():
    print(3234)
    with open("index_data.html", "r") as file:
        html_content = file.read()

    return HTMLResponse(content=html_content, status_code=200)


if __name__ == "__main__":
    import uvicorn

    app.mount('/', flet_fastapi.app(root_main))
    uvicorn.run(app, host="0.0.0.0", port=25022)


HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>
    <script src="https://telegram.org/js/telegram-web-app.js"></script>
    <script>
        let tg = window.Telegram.WebApp;

        let user_id = tg.initDataUnsafe.user.username;
        let colorScheme = tg.colorScheme;

        // Отправка name и colorScheme на сервер FastAPI
        fetch('/get_data', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                user_id: user_id,
                colorScheme: colorScheme
            })
        })
        .then(response => response.json())
        .then(data => {
            console.log('Success:', data);
        })
        .catch((error) => {
            console.error('Error:', error);
        });

        // Вызов метода close()

    </script>
</body>
</html>
```

### ------------------------------------------------------

- [X] I have searched for answers to my question both in the issues and in previous discussions.

**Comment by tanmay-bhatgare** [comment]
Honestly, I think it's better to keep `flet_fastapi` and `flet` separate for now. `felt_fastapi` is still pretty new and doesn’t have a big community yet, so mixing them might cause issues later. Separating the backend (FastAPI) and frontend (Flet) could save you time and headaches. I’m working on a task manager app with FastAPI and Flet, and this approach has been working well for me. If you have further ideas to make it work please share.

**Reply by RRozi** [reply]
Well, could you please send me an example to solve my problem?

**Reply by tanmay-bhatgare** [reply]
I do like to but what should I send? A simple integration of flet app with FastAPI backend or something else?

**Reply by RRozi** [reply]
Yes, I need to receive data from the one I specified when I exited and pass it to the frontend

**Reply by tanmay-bhatgare** [reply]
Can you share entire codebase in your GitHub profile so I can help you?

**Reply by RRozi** [reply]
Весь код проекта выше

**Reply by tanmay-bhatgare** [reply]
Then what are imports of gui and api? These aren't python standard libraries, are they?

**Reply by RRozi** [reply]
And these are unnecessary libraries that I forgot to remove, I tried to use a different approach

**Reply by tanmay-bhatgare** [reply]
hey can you explain the steps to reproduce this code i am getting lil bit confuse as I am not familiar with flet_fastapi and telegram api?

**Reply by RRozi** [reply]
I just need to be able to Flat execute telegram api and return received data to flat page

**Reply by tanmay-bhatgare** [reply]
Have you tried this without flet? You should check why isn't working, one solution would be ChatGPT.

**Reply by RRozi** [reply]
ChatGPT It won't help. How can I do it without flet if my app is on flet?

**Reply by tanmay-bhatgare** [reply]
No I mean your using flet as gui(frontend) so did you just tried only `fastapi` and `html`. Once if everything worked you can tell implement it in flet or I can help you.

**Reply by RRozi** [reply]
I would be grateful if you could help me

**Reply by RRozi** [reply]
Telegram also gives the ability to work with the back button, here is an example of the code
var BackButton = WebApp.BackButton;
BackButton.show();
BackButton.onClick(function() {
  WebApp.showAlert("Нет пути назад!");
  BackButton.hide();
});
WebApp.onEvent('backButtonClicked', function() {
  /* CODE */
});
And just when you click and process the layout, the flet should go to another page. page.go("/")
And so that these are not crutches, but everything is clear, when you click on the button through Fast API, the page switching method takes a long time to be called

**Reply by tanmay-bhatgare** [reply]
Mate sorry but without knowing how to reproduce code i can't help. I am getting this message whenever i try to run index_data.html file.

```javascript
[Telegram.WebView] > postEvent web_app_set_header_color Object
telegram-web-app.js:135 [Telegram.WebView] > postEvent web_app_set_bottom_bar_color Object
telegram-web-app.js:135 [Telegram.WebView] > postEvent web_app_request_theme 
telegram-web-app.js:135 [Telegram.WebView] > postEvent web_app_request_viewport 
(index):12 Uncaught TypeError: Cannot read properties of undefined (reading 'username')
    at (index):12:44
``` 
you have any idea about it?

**Reply by RRozi** [reply]
There is no need to launch it, it serves simply as an appeal to api in the skin at the very beginning you can get acquainted with how it should work

**Reply by tanmay-bhatgare** [reply]
mate look let's restart. I tried to run your code but fastapi server get served but flet app didn't load, idk why can you please tell me how to reproduce this code, this might help you too. Thank!

**Reply by RRozi** [reply]
Bro look, that's the joke, I can get data but the app won't show up or vice versa, I can get the app but not get the data. Right now I'm trying from the flat app directly access the API, and here too there is a problem, the first endpoint that performs html It is executed successfully, but already in the html Where it returns data to another endpoint, it doesn't work at all.

**Reply by tanmay-bhatgare** [reply]
where are you getting data in browser console or terminal where you ran your program? and about the second endpoint `/data` you're returning a `HTMLResponse`, this will return a html in browser, means if there's a javascript code in html it won't work unless you render `.html` file in your browser. So if you want your `data(user_id and color_scheme)` to be displayed, you might have to find a different way like telegram api which can send you data directly without javascript included. In short you can't use HTML rendering code(includes javascript) fastapi route in flet app. This i entire conclusion I came up with.
